File: algo/src/test/java/com/hartwig/actin/algo/evaluation/washout/WashoutTestFactory.java
package com.hartwig.actin.algo.evaluation.washout;

import java.time.LocalDate;
import java.util.List;

import com.hartwig.actin.ImmutablePatientRecord;
import com.hartwig.actin.PatientRecord;
import com.hartwig.actin.TestDataFactory;
import com.hartwig.actin.algo.medication.MedicationStatusInterpretation;
import com.hartwig.actin.algo.medication.MedicationStatusInterpreter;
import com.hartwig.actin.clinical.datamodel.ImmutableClinicalRecord;
import com.hartwig.actin.clinical.datamodel.ImmutableMedication;
import com.hartwig.actin.clinical.datamodel.Medication;
import com.hartwig.actin.clinical.datamodel.MedicationStatus;
import com.hartwig.actin.clinical.datamodel.TestClinicalFactory;

import org.apache.logging.log4j.util.Strings;
import org.jetbrains.annotations.NotNull;

final class WashoutTestFactory {

    private WashoutTestFactory() {
    }

    @NotNull
    public static MedicationStatusInterpreter activeFromDate(@NotNull LocalDate referenceDate) {
        return medication -> {
            LocalDate stopDate = medication.stopDate();
            if (stopDate == null || !referenceDate.isAfter(stopDate)) {
                return MedicationStatusInterpretation.ACTIVE;
            } else {
                return MedicationStatusInterpretation.UNKNOWN;
            }
        };
    }

    @NotNull
    public static ImmutableMedication.Builder builder() {
        return ImmutableMedication.builder().name(Strings.EMPTY).codeATC(Strings.EMPTY).status(MedicationStatus.ACTIVE);
    }

    @NotNull
    public static PatientRecord withMedications(@NotNull List<Medication> medications) {
        return ImmutablePatientRecord.builder()
                .from(TestDataFactory.createMinimalTestPatientRecord())
                .clinical(ImmutableClinicalRecord.builder()
                        .from(TestClinicalFactory.createMinimalTestClinicalRecord())
                        .medications(medications)
                        .build())
                .build();
    }
}

File: algo/src/test/kotlin/com/hartwig/actin/algo/evaluation/washout/WashoutTestFactory.kt
package com.hartwig.actin.algo.evaluation.washout

import com.hartwig.actin.clinical.interpretation.MedicationStatusInterpretation
import com.hartwig.actin.clinical.interpretation.MedicationStatusInterpreter
import com.hartwig.actin.datamodel.PatientRecord
import com.hartwig.actin.datamodel.TestPatientFactory
import com.hartwig.actin.datamodel.clinical.AtcClassification
import com.hartwig.actin.datamodel.clinical.Medication
import com.hartwig.actin.datamodel.clinical.MedicationStatus
import com.hartwig.actin.datamodel.clinical.TestMedicationFactory
import java.time.LocalDate

internal object WashoutTestFactory {

    private val base = TestPatientFactory.createMinimalTestWGSPatientRecord()

    fun activeFromDate(referenceDate: LocalDate): MedicationStatusInterpreter {
        return object : MedicationStatusInterpreter {
            override fun interpret(medication: Medication): MedicationStatusInterpretation {
                val stopDate = medication.stopDate
                return if (stopDate == null || !referenceDate.isAfter(stopDate)) {
                    MedicationStatusInterpretation.ACTIVE
                } else {
                    MedicationStatusInterpretation.UNKNOWN
                }
            }
        }
    }

    fun medication(
        atc: AtcClassification? = null,
        stopDate: LocalDate? = null,
        name: String = "",
        isTrialMedication: Boolean = false
    ): Medication {
        return TestMedicationFactory.createMinimal().copy(
            atc = atc,
            stopDate = stopDate,
            name = name,
            isTrialMedication = isTrialMedication,
            status = MedicationStatus.ACTIVE
        )
    }

    fun withMedications(medications: List<Medication>): PatientRecord {
        return base.copy(medications = medications)
    }
}